Add Open Policy Agent material (#607)
trinodb/trino.io: _posts/2024-02-04-opa-arrives.md

---
layout: post
title:  "Open Policy Agent for Trino arrived"
author: Manfred Moser
image: /assets/images/logos/opa-small.png
excerpt_separator: <!--more-->
show_pagenav: true
---

Trino now ships with an access control integration using the popular and widely
used [Open Policy Agent (OPA)](https://www.openpolicyagent.org/) from the Cloud Native
Computing Foundation. The release of [Trino
438]({{site.url}}/docs/current/release/release-438.html) marks an important
milestone of the effort towards this integration.

<!--more-->

## Collaboration and history

Open Policy Agent was first released in 2016 and has gained more and more

> popularity in the ecosystem of cloud native application and beyond.

Reviewer: ```suggestion
popularity in the ecosystem of cloud native applications and beyond.
```

Initial efforts for an integration with Trino started at Bloomberg, Stackable,

> Raft and others places separately, in parallel and partially in collaboration.

Reviewer: ```suggestion
Raft and others places separately and sometimes in parallel, with only partial collaboration.
```

You first heard about it more in August 2022 in the [Trino Community Broadcast
episode 39]({{site.url}}/episodes/39.html) with a team from Raft as guests.

> Usage and experience with OPA grew and in the end Pablo Arteaga from
> [Bloomberg](https://www.bloomberg.com) and Sönke Liebau from

Reviewer: ```suggestion
Usage and experience with OPA grew. In the end, Pablo Arteaga from
[Bloomberg](https://www.techatbloomberg.com/) and Sönke Liebau from
```

Minor reword? And could we use https://www.techatbloomberg.com/ instead of Bloomberg.com?

to Trino. Their persistence and collaboration led them through many review
comments, update commits, and even a second PR, to submit a talk and eventually
present about the Open Policy Agent access control with Trino and their
motivation to move from Apache Ranger to OPA.

## OPA at Trino Summit 2023

The presentation from Pablo and Sönke titled "Trino OPA authorizer - An open
source love story" received a lot of interest from the audience at the event and
on YouTube since then:

 {% youtube fbqqapQbAv0 %}

They also made the [slide deck available for your
reference]({{site.url}}/assets/blog/trino-summit-2023/opa-trino.pdf).

Edward Morgan and Bhaarat Sharma from [Raft](https://teamraft.com/) also
presented [Avoiding pitfalls with query federation in data
lakehouses](https://www.youtube.com/watch?v=6KspMwCbOfI) at Trino Summit, and
detailed their OPA usage.

 {% youtube 6KspMwCbOfI %}

## OPA on the way to Trino

Pablo and Sönke continued their efforts on the [pull
request](https://github.com/trinodb/trino/pull/19532) after Trino Summit. They
worked successfully with Dain on the code review and necessary changes, and
helped Manfred with the documentation.

> Finally, with the release of Trino 438 the [Open Policy Agent access
> control]({{site.url}}docs/current/security/opa-access-control.html) is available

Reviewer: ```suggestion
Finally, with the release of Trino 438, the [Open Policy Agent access
control]({{site.url}}docs/current/security/opa-access-control.html) is available
```

to all Trino users.

The community is already taking notice with follow up pull requests for further
improvements and blog posts such as [Enhancing Security and Observability in
Trino with Open Policy Agent and
OpenTelemetry](https://www.linkedin.com/pulse/enhancing-security-observability-trino-open-policy-agent-isa-inalcik-zhl9e/)
from Isa Inalcik.

> ## What next

Author: Updated

> We reached another milestone but we are not done yet. Specifically for OPA we

Reviewer: ```suggestion
We reached another milestone but we are not done yet. Specifically for OPA, we
```

are looking at the following next tasks:

* Get more features from various older, private forks converted into pull
  requests to Trino so everyone can benefit.
* Update the documentation with more practical advice and tips.
* Provide further resources for running OPA with Trino, writing rego scripts,
  and helping the community.

> 

Reviewer: ```suggestion
* Implementation of Row Level Filtering and Column Masking, a [draft](https://github.com/bloomberg/trino/pull/16) is already underway
```

We could mention filtering and masking?
